Auto name variants and manual name corrections

davidweichiang opened the pull request:
bin/auto_name_variants is a script for automatically detecting name variants.

The initial run showed a lot of errors, so these were corrected before attempting to build a list of name variants.

I corrected a lot of cases where the first and last names had obviously been switched, but there were many, many cases that I left alone. The auto name variants usually group them pretty well.

There were many cases of names in all caps; I didn't change any of these. I think they could be changed automatically, and I think they probably should.

mbollmann commented on bin/auto_name_variants.py:
```diff
@@ -0,0 +1,185 @@
+            print()
+            print('- canonical: {{first: {}, last: {}}}{}'.format(cfirst, clast, comment))
+        if len(names) > 1:
+            print('- variants:')
```
That isn't the right output, as it produces a new list entry with just "variants" in it. However, I would *strongly* recommend to do this with a YAML library instead anyway...

davidweichiang commented:
Oh, sorry. I'll redo with a YAML library.

davidweichiang commented:
In going through the name variants, I see that a lot of them vary only in casing or accents. Should we enumerate these explicitly in the name_variants.yaml file or should the code automatically fold these together? (Or does it do that already anyway?) How about camel-casing vs. spacing, as in "vanBeethoven" vs. "van Beethoven"?

mbollmann commented:
I would not suggest conflating anything automatically. It would make resolving name collisions (#208) more difficult, and it would require new mechanisms for cases where this behaviour *isn't* wanted. For example, "Muller" and "Müller" are both actual surnames.

mbollmann commented:
A nice example I just found: https://www.aclweb.org/anthology/L18-1039 has both "Marcos García Salido" and "Marcos Garcia" on the author list. Apart from the second surname that distinguishes them here, their names also differ in that one uses a diacritic on the í while the other does not.

davidweichiang commented:
OK, I've updated the script to use PyYAML and regenerated. Since PyYAML doesn't generate comments, I've added a new field called `papers` that is intended to be deleted eventually. But it's very useful for checking whether a cluster of variants is good.

davidweichiang commented:
The build is failing because I've written a lot of author names as

<author><last>Chiang</last><first>Wei</first></author>

This is because that's the order the names are written in in the original paper.

Do you prefer that I fix these, or that we extend the schema to allow the names in either order?

mbollmann commented:
I don't see the `name_variants.auto.yaml` actually having been updated?

About the first name/last name thing, I'm pretty confident that my scripts don't care about the actual order of these elements. There might be more consequences to this that I'm not aware of though, and maybe it's also a question of principle. @mjpost?

mjpost commented:
I think my principle on XML schemas is "as tight as possible, as free as necessary". This author order can be fixed with a Perl one-line so I say we keep the schema. As a consolation I will do it myself a minute now...

mjpost commented:
Let's see if that works...

mjpost commented:
Ah, that failed to catch lines with `<author>` entries split across lines.

davidweichiang commented:
Oops, I will add the auto output soon.

I was too slow to respond about the first-last ordering. I think it's a feature, not a bug, to preserve the ordering as the author chose to write it in the original paper.

Swapped first/last names are probably the most common error I had to correct in the XML files, which to me indicates that people who come from family-name-first cultures don't easily adapt, or don't want to adapt, to a family-name-last convention.

To me, that suggests that it would be good to preserve the original ordering information, in case, for example, the Anthology decided to display author names in the author's preferred order.

mjpost commented:
Okay, feel free to revert the changes I made. I don't have a strong preference here, but I don't think it's true in general that the components of an author's name in our XML mirror what was in the input bib or in the paper, right? My understanding is that `first` and `last` correspond to what goes after and before the comma in the output BibTeX.

davidweichiang commented:
You're right on both counts. There are lots of places where the input .bib and the .xml diverge from what's actually written on the paper. And it's definitely the case (as long as we have BibTeX) that we want the BibTeX to have `author = "Lastname, Firstname"`.

But, I guess the principle I'm following is that preserving information is better than discarding information, so having some entries that preserve the original name ordering is better than having no entries that preserve it.

If you're okay with rolling back https://github.com/acl-org/acl-anthology/pull/217/commits/b4703a8c0605d1f5adcb0886a669224a633d544c -- is there an easy git command that will do that?

davidweichiang commented:
The build is definitely still failing on the remaining cases where the last name is first (but are on separate lines so the Perl script didn't catch them).

mjpost commented:
I am still confused. `first` and `last` already have a semantics—or maybe it's better to say (given [the general chaos around names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/)) they have a role—that determines what goes before or after the comma in the bibtex. But I am not convinced that imbuing the *order* of those tags in the XML with further meaning is a good idea, particularly when we don't always preserve that order, and have no way of indicating when we've preserved it and when we haven't.

Maybe I'm just confused. For your example:

    <author><last>Chiang</last><first>Wei</first></author>

I assume you imply the following:

1. The author put "Chiang Wei" in the paper PDF
2. "Chiang" is actually what we'd consider the surname, and therefore

When we generate the BibTeX, I further assume that we will generate "Chiang, Wei", because I assume the code uses the tag values and not their order. Is this true, @mbollmann? Is this what you intend, @davidweichiang?

davidweichiang commented:
Yes, that's basically what I have in mind. (A problem for my view is if I wrote my name as "David Chiang Wei", in which case my given name, "David Wei," is not even a contiguous substring. But I've only seen this once.)

mjpost commented:
Okay, I understand now. I am still worried that this is an implicit semantics that is going to get lost. To anyone reading the file and not in the know, it will just look like the `<first>` and `<last>` tags were inadvertently inverted.

mjpost commented:
Another option would be to keep the current schema, and add the original display order as a "displayOrder" attribute.

mbollmann commented:
> When we generate the BibTeX, I further assume that we will generate "Chiang, Wei", because I assume the code uses the tag values and not their order. Is this true, @mbollmann?

[Yes.](https://github.com/acl-org/acl-anthology/blob/master/bin/anthology/people.py#L70)

davidweichiang commented:
I guess it depends on if you view an XML tag as a key-value store or as rich text -- I tend towards the view that it's rich text, in which case it's natural to put the fields in their original order.

Admittedly, it would be too hard / too strict to follow this view 100%, but, as I tried to communicate above but maybe not very clearly, my sense from looking through a lot of papers is that many people (understandably) do consistently prefer their native-language order, and it would be nice not to throw away information that might in the future help to accommodate that preference.

mjpost commented:
I am with you entirely on the idea of preserving this information—as well as the view of XML as rich text. My only concern was that last-first wouldn't register as meaningful, since people aren't necessarily primed to place meaning on the order of these two particular keys (as they are, for example, for author order).

Let's go with your suggestion to expand the schema to allow this. It seems no one else shares this concern, so if you tag this conversation in the git commit, that may serve as sufficient documentation. We might also add a comment in the schema.

davidweichiang commented:
Thanks for your flexibility! Let's see if this works.

mjpost commented:
Great, we're passing now. I've spot-checked and haven't seen any problems. Do you think it's good to merge?

davidweichiang commented:
I think so. There will undoubtedly be more corrections to author names, but I see that master also has some corrections so maybe it's good to stay in sync.

davidweichiang commented:
And assuming you're okay with a temporary file `name_variants.auto.yaml` floating around in master.

mjpost commented:
None of this is currently used, though—do you mean for us to incorporate `name_variants.auto.yaml` into the author page generation? I think that would require the file to be read in `bin/anthology/people.py`.

davidweichiang commented:
No, there are some false positives in name_variants.auto.yaml so I don't think it should be used directly. I plan to check them and move the good ones into the real name_variants.yaml. So on second thought, let's leave this unmerged until that's done. If anyone wants to help check, they are welcome to.

mjpost commented:
Okay, that makes sense.

BTW, I see you are adding the subset of papers associated with each name token string. We haven't discussed this---and possibly it is your intent---but this approach is a potential solution to the name collision problem (#208).

mjpost commented:
I added  "[WIP]" (work in progress) flag to the title to note that we're not ready to merge this.

davidweichiang commented:
Where would be an appropriate place to put information (like affiliations) for people with very similar names? Can a comment field be added to name_variants.yaml? (Actual comments aren't convenient because some YAML processors strip them out.) Or a separate text file?

davidweichiang commented:
Will something bad happen if name_variants has:

```
- canonical: {first: Robert L., last: Mercer}
  variants:
  - {first: Robert, last: Mercer}
- canonical: {first: Robert E., last: Mercer}
  - {first: Robert, last: Mercer}
```

Because that unfortunately happened. Unlike the case of Yang Liu and Yang Liu, who always publish (in English) under the same name and so have coextensive entries until we can figure out issue #208, the two Robert Mercers usually published with their middle initials, but both sometimes published without.

mbollmann commented:
> Will something bad happen if name_variants has:

If this doesn't error out yet, it definitely should. There is no way to decide which of the two entries an author named "Robert Mercer" should be associated with, so I don't even know how this *could* be handled at all.

> Where would be an appropriate place to put information (like affiliations) for people with very similar names?

Affiliation can change over time and from paper to paper, so I'm not sure it makes sense to associate it globally with an author entry. Same goes for e-mail. Unless we'd have this information on a paper-by-paper basis, I'm not sure what we could do with it?

davidweichiang commented:
For the Robert Mercer case, do you have any recommendations for how to write this in name_variants.yaml for now?

Regarding extra information, what I have in mind is not something that would appear on the website, but just comments that might be helpful to someone updating name_variants.yaml in the future. For example, if the entry for Robert L. Mercer says "IBM, Renaissance" and Robert E. Mercer says "Western Ontario," that would be helpful to someone later trying to disambiguate the papers by Robert Mercer.

mbollmann commented:
> For the Robert Mercer case, do you have any recommendations for how to write this in name_variants.yaml for now?

I don't really see any way to represent this right now, unfortunately. I think it's essentially the same problem as other name collisions. The best we could do is merge these entries for now and maybe change the "Also published as..." on the website to "Includes entries for...", so it doesn't necessarily imply all papers are by the same person.

We could also add comments to find these cases again easily later. If YAML comments are inconvenient, I think a `comment:` entry in the dict should be totally fine.

Anything that actually distinguishes the two would require implementing a solution for #208 first, IMHO.

davidweichiang commented:
Thanks. Is it still an error to include entries that have no variants? (How about an empty variants list?)

mbollmann commented:
Not sure what the current behavior is, but if it's deemed useful I can make the code handle it, of course.

davidweichiang commented:
How does that look?

One concern is some names like "A. Zhang" where the last name is a very common last name. In some cases there was a collision, so I just left it off; in some cases, there was no collision, so I grouped it with the canonical name it belonged to. But there's a decent chance that in the future there will be a collision.

mjpost commented:
I looked through it a bit and the entries looked pretty good.

What do the square brackets mean here? I couldn't find the raw text anywhere.

```
- canonical: {first: Yao-zhong, last: Zhang}
  variants:
  - {first: Yao Zhong, last: Zhang}
  - {first: 'Y[ao-Zhong]', last: Zhang}
```

mjpost commented:
There are definitely a lot of errors that this is turning up, e.g., search for the "van" section (Marieke Van Erp, Josef Van Genabith, many others).

mjpost commented:
I wonder if we should add a "report error" button in the "AKA" line on the web page, linking to a templated issue that fills in the name, making it easy to manually correct.

davidweichiang commented:
@mjpost Good job finding the one quirk in a 14,000 line file. :) That was where I found an arXiv version of that paper, and the arXiv metadata had full names. Since I was 100% sure of the author's full name, I wrote its completion in square brackets in the XML. If you want to get rid of that, or format it differently, it's easy to change because it only happened once. (But could happen much more if an effort was made to confirm full names of authors...)

I didn't make much attempt to normalize many classes of variations, like "van"s or Spanish double surnames or Romanian comma-belows. There are too many invidual preferences for me to guess which variations are errors. With "van," some people might want it in the first name (so it sorts by "Erp" or "Genabith") and some people might want it in the last name (so it sorts by "van"); some people capitalize "van" and some people don't capitalize the word after ("hamme"), and some (Americans) fuse the "van" into the last name ("Vanderwende").

davidweichiang commented:
It's now failing because there's no author page for `raman-chandrasekar`. This is because I made "Raman Chandrasekar" the canonical spelling of his name (https://www.khoury.northeastern.edu/people/raman-chandrasekar/) even though it's never spelled that way in any Anthology paper. (It's spelled `R. Chandrasekar` and `Raman Chandraseker`).

@mbollmann If it's a real pain to modify the code to handle this case, I can make one of the actual spellings the canonical one, but neither of them is a great choice.

mjpost commented:
Why is neither a great choice? The "Chandraseker" spelling is the in PDF, suggesting that that spelling was okay with him (unless a co-author bungled it?): https://aclweb.org/anthology/M98-1022.

davidweichiang commented:
It wouldn't be the first time that an author's name is misspelled on a paper (see @MartiHearst's comment https://github.com/acl-org/acl-anthology/issues/208#issuecomment-475913997). I know the author and don't recall ever seeing it spelled this way, and a Google search for this spelling turns up only references to this paper.

mjpost commented:
I suggest we fix the XML here, then, instead of complexifying the code for a single instance. It goes against the principle of having metadata reflect what's on the paper, but it seems okay to me for a correction.

davidweichiang commented:
I think that one-line fix worked? It's passing now, anyway.

davidweichiang commented:
Closes #86, although I don't know how to keep name_variants.yaml up to date going forward. Hopefully, since most meetings are using START, author names will be more consistent.

mjpost commented:
I think this is not true: We host a number of non-ACL conferences, and most of these (anecdotally) seem to use easy chair. LREC, for example, is a big one, IJCAI is also interested in sending us their data. I hope this continues to grow, and it means that we should probably focus efforts on sanitizing data that comes in. I think that we will have to add running this script to the ingestion pipeline.

davidweichiang commented:
Or maybe we should rely on Semantic Scholar to maintain name variants as well as collisions?

MartiHearst commented:
We could at least explore what it would be like to work with Semantic Scholar.  There are also some other groups doing universal naming; there is the Springer Nature SciGraph tool and they have an open source effort.  My contact there is Henning Schoenenberge.  However, my inclination is to start with SS as their entire enterprise is open access and they have an AI focus.

davidweichiang commented:
In any case, I believe that this PR is ready to merge -- I think we can continue discussing future plans in #208.
